### PR TITLE
[Gao Haochun] iP

### PR DESCRIPTION
**Pull request by Harry-Gao-H**
## Duke 10 Plus

> We're living the future so
> the present is our past.

Duke 10 Plus frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* SUPER FAST to use

All you need to do is,
1. download it from [here](www.google.com).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method:
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

### CODE REVIEW

File: src/main/java/duke/Task.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    Task(String content, boolean isDone) {
+        this.content = content;
+        this.isDone = isDone;
+    }
+
+    void setDone() {
+        this.isDone = true;
+    }
+
+    String encoding() {
```
Comment (LMAOboxhack) [inline src/main/java/duke/Task.java]:
```suggestion
    String encode() {
```
Would it be better to use the verb version of the word instead, as outlined in the [Code Quality: Naming](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/basic/nounsAndVerbsAsNames/) section of the textbook?

File: src/main/java/duke/Parser.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+     * Returns error messages or a string showing the added task.
+     * The input string must follow input format, otherwise error messages will be return.
+     * The new task will be created and added to lst.
+     * @param input input message
+     * @param taskList the taskList containing all tasks
+     * @return a string showing the added task and number of tasks
+     */
+     static String addList(String input, TaskList taskList) {
+         ArrayList<Task> lst = taskList.getTasks();
+         ArrayList<String> validType = new ArrayList<>(
```
Comment (LMAOboxhack) [inline src/main/java/duke/Parser.java]:
```suggestion
         ArrayList<Task> tasks = taskList.getTasks();
         ArrayList<String> validTypes = new ArrayList<>(
```
Would it be better to name your variables in plural form instead, to show that they are collections of items? See [this](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/basic/nounsAndVerbsAsNames/) section of the textbook.

File: src/main/java/duke/Deadline.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+class Deadline extends Task {
+    protected String by;
```
Comment (LMAOboxhack) [inline src/main/java/duke/Deadline.java]: Would it be good to use a more descriptive variable name here? Right now "by" is not very indicative of the variable's function in the class. See '[Use name to explain](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/intermediate/useNameExplain/)' from the textbook.

File: src/main/java/duke/Parser.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+    /**
+     * Returns the message that shows which task is marked as completed.
+     * Mark the task indicated by index as completed.
+     * @param index the index of task to be marked
+     * @param taskList the taskList containing all tasks
+     * @return the messages
+     */
+    static String doTask(String index, TaskList taskList) {
+        int idx;
```
Comment (LMAOboxhack) [inline src/main/java/duke/Parser.java]:
```suggestion
        int indexToMarkComplete;
```
Would it be better to use standard words in your variable naming here? Also might be good to specify a more descriptive variable name. See '[Use standard words](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/basic/useStandardWords/)' & '[Use name to explain](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/intermediate/useNameExplain/)' in the textbook.

File: src/main/java/duke/Storage.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Scanner;
+
+public class Storage {
+    String filePath;
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    public ArrayList<Task> loadTasks() throws DukeException {
+        File f = new File(filePath);
```
Comment (LMAOboxhack) [inline src/main/java/duke/Storage.java]:
```suggestion
        File fileToLoad = new File(filePath);
```

Would it be better to specify a more descriptive variable name here? See '[Use standard words](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/basic/useStandardWords/)' & '[Use name to explain](https://nus-cs2103-ay2122s1.github.io/website/book/codeQuality/nameWell/intermediate/useNameExplain/)' in the textbook.

**Review by LMAOboxhack — COMMENTED** [review]
Looks good 👍 just some minor issues such as method and variable naming.

**Review by mukundrs — COMMENTED** [review]
Your code looks really good to me! Just a few nitpicks here and there but nothing too imp. Some classes and functions are missing javadocs. While they do seem redundant sometimes, its generally good practice to include them anyways. There were a few lines of commented out code which personally I would remove so that it improves readability. Apart from those nitpicks  I thought your code was very easy to read and very well done! 😄

File: src/main/java/DialogBox.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * An example of a custom control using FXML.
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
+ */
+public class DialogBox extends HBox {
+    @FXML
+    private Label dialog;
+    @FXML
+    private ImageView displayPicture;
+
```
Comment (mukundrs) [inline src/main/java/DialogBox.java]: Missing javadocs for constructor

File: src/main/java/DialogBox.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
+        Collections.reverse(tmp);
+        getChildren().setAll(tmp);
+        setAlignment(Pos.TOP_LEFT);
+    }
+
```
Comment (mukundrs) [inline src/main/java/DialogBox.java]: Missing javadocs for function

File: src/main/java/DialogBox.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
```
Comment (mukundrs) [inline src/main/java/DialogBox.java]: Missing javadocs for functions

File: src/main/java/Duke.java [hunk 9]
```diff
@@ -1,10 +1,197 @@
...
+//import javafx.scene.image.Image;
+//import javafx.scene.image.ImageView;
+//import javafx.stage.Stage;
+
+import duke.DukeException;
+import duke.Storage;
+import duke.TaskList;
+import duke.Ui;
+import javafx.fxml.FXML;
+
```
Comment (mukundrs) [inline src/main/java/Duke.java]: Missing javadocs for class description

File: src/main/java/Duke.java [hunk 10]
```diff
@@ -1,10 +1,197 @@
...
+        ui = new Ui();
+        storage = new Storage(FILEPATH);
+        try {
+            taskList = new TaskList(storage.loadTasks());
+        } catch (DukeException e) {
+            ui.showLoadingError(e.toString());
+            taskList = new TaskList();
+        }
+    }
+
```
Comment (mukundrs) [inline src/main/java/Duke.java]: Missing javadocs for function

File: src/main/java/Duke.java [hunk 11]
```diff
@@ -1,10 +1,197 @@
...
+//     */
+//    private void handleUserInput() {
+//        Label userText = new Label(userInput.getText());
+//        Label dukeText = new Label(getResponse(userInput.getText()));
+//        dialogContainer.getChildren().addAll(
+//                DialogBox.getUserDialog(userText, new ImageView(user)),
+//                DialogBox.getDukeDialog(dukeText, new ImageView(duke))
+//        );
+//        userInput.clear();
+//    }
```
Comment (mukundrs) [inline src/main/java/Duke.java]: Would personally remove these comments to improve readability

File: src/main/java/Duke.java [hunk 12]
```diff
@@ -1,10 +1,197 @@
...
+    Ui getUi() {
```
Comment (mukundrs) [inline src/main/java/Duke.java]: Missing javadocs for function

File: src/main/java/Main.java [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+import javafx.scene.layout.AnchorPane;
+import javafx.stage.Stage;
+
+/**
+ * A GUI for Duke using FXML.
+ */
+public class Main extends Application {
+
+    private Duke duke = new Duke();
+
```
Comment (mukundrs) [inline src/main/java/Main.java:17]: Missing javadocs for function

File: src/main/java/MainWindow.java [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    @FXML
+    private TextField userInput;
+    @FXML
+    private Button sendButton;
+
+    private Duke duke;
+
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
+    private Image dukeImage = new Image(this.getClass().getResourceAsStream("/images/DaDuke.png"));
+
```
Comment (mukundrs) [inline src/main/java/MainWindow.java]: Missing javadocs

File: src/main/java/MainWindow.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+    @FXML
+    public void initialize() {
+        String output = "     Hello! I'm Duke\n"
+                + "     What can I do for you?\n";
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+        dialogContainer.getChildren().addAll(
+                DialogBox.getDukeDialog(output, dukeImage)
+        );
+    }
+
```
Comment (mukundrs) [inline src/main/java/MainWindow.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+class Deadline extends Task {
+    protected String by;
+    protected LocalDate date;
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    Deadline(String content, String by) {
+        super(content);
+        this.by = by;
+        try {
+            date = LocalDate.parse(by);
+        } catch (DateTimeParseException e) {
+            date = null;
+        }
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+    Deadline(String content, String by, boolean isDone) {
+        super(content, isDone);
+        this.by = by;
+        try {
+            date = LocalDate.parse(by);
+        } catch (DateTimeParseException e) {
+            date = null;
+        }
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    LocalDate getDate() {
+        return date;
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Deadline.java [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+    @Override
+    String encoding() {
+        return "D&&" + super.encoding() + "&&" + by;
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Deadline.java]: Missing javadocs

File: src/main/java/duke/Event.java [hunk 22]
```diff
@@ -0,0 +1,51 @@
+package duke;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
```
Comment (mukundrs) [inline src/main/java/duke/Event.java]: Missing javadocs

File: src/main/java/duke/Event.java [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+class Event extends Task {
+    protected String at;
+    protected LocalDate date;
+
```
Comment (mukundrs) [inline src/main/java/duke/Event.java]: Missing javadocs

File: src/main/java/duke/Event.java [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+    Event(String content, String at) {
+        super(content);
+        this.at = at;
+        try {
+            date = LocalDate.parse(at);
+        } catch (DateTimeParseException e) {
+            date = null;
+        }
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Event.java]: Missing javadocs

File: src/main/java/duke/Event.java [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+    public String toString() {
+        String time;
+        if (date == null) {
+            time = at;
+        } else {
+            time = date.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+        }
+        return "[E]" + super.toString() + " (at: " + time + ")";
+    }
+}
```
Comment (mukundrs) [inline src/main/java/duke/Event.java]: Missing javadocs

File: src/main/java/duke/Parser.java [hunk 26]
```diff
@@ -0,0 +1,221 @@
+package duke;
+
+import java.time.LocalDate;
+import java.util.ArrayList;
+import java.util.Arrays;
+
```
Comment (mukundrs) [inline src/main/java/duke/Parser.java]: Missing javadocs

File: src/main/java/duke/Parser.java [hunk 27]
```diff
@@ -0,0 +1,221 @@
...
+            return new Todo(strings[2], strings[1].equals("Done"));
+        case "D":
+            return new Deadline(strings[2], strings[3], strings[1].equals("Done"));
+        case "E":
+            return new Event(strings[2], strings[3], strings[1].equals("Done"));
+        default:
+            return null;
+        }
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Parser.java]: Missing javadocs

File: src/main/java/duke/Parser.java [hunk 28]
```diff
@@ -0,0 +1,221 @@
...
+        s.append("    Nice! I've marked this task as done: \n");
+
+        tasks.get(idx - 1).setDone();
+        s.append("       " + tasks.get(idx - 1).toString() + "\n");
+
+        s.append("    ____________________________________________________________\n");
+
+        return s.toString();
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Parser.java]: Missing javadocs

File: src/main/java/duke/Storage.java [hunk 29]
```diff
@@ -0,0 +1,58 @@
+package duke;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
Comment (mukundrs) [inline src/main/java/duke/Storage.java]: Missing javadocs

File: src/main/java/duke/Task.java [hunk 30]
```diff
@@ -0,0 +1,40 @@
...
+        return null;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[%s] %s", isDone ? "X" : " ", this.content);
+    }
+
+
+}
```
Comment (mukundrs) [inline src/main/java/duke/Task.java]: Missing javadocs

File: src/main/java/duke/TaskList.java [hunk 31]
```diff
@@ -0,0 +1,36 @@
+package duke;
+
+import java.util.ArrayList;
+
```
Comment (mukundrs) [inline src/main/java/duke/TaskList.java]: Missing javadocs

File: src/main/java/duke/TaskList.java [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+public class TaskList {
+    private ArrayList<Task> tasks;
+    public TaskList() {
```
Comment (mukundrs) [inline src/main/java/duke/TaskList.java]: Missing javadocs and would personally have an empty line in between tasks and tasklist()

File: src/main/java/duke/TaskList.java [hunk 33]
```diff
@@ -0,0 +1,36 @@
...
+        } catch (IndexOutOfBoundsException e) {
+            throw new DukeException("The task does not exit");
+        }
+    }
+
+    void addTask(Task t) {
+        tasks.add(t);
+    }
+
+}
```
Comment (mukundrs) [inline src/main/java/duke/TaskList.java]: Missing javadocs

File: src/main/java/duke/Todo.java [hunk 34]
```diff
@@ -0,0 +1,21 @@
...
+    @Override
+    String encoding() {
+        return "T&&" + super.encoding();
+    }
+
+    @Override
+    public String toString() {
+        return "[T]" + super.toString();
+    }
+}
```
Comment (mukundrs) [inline src/main/java/duke/Todo.java]: Missing javadocs

File: src/main/java/duke/Ui.java [hunk 35]
```diff
@@ -0,0 +1,70 @@
+package duke;
+
+import java.util.ArrayList;
+import java.util.Scanner;
+
```
Comment (mukundrs) [inline src/main/java/duke/Ui.java]: Missing javadocs

File: src/main/java/duke/Ui.java [hunk 36]
```diff
@@ -0,0 +1,70 @@
...
+     * Displays the greeting message.
+     */
+    public String showGreeting() {
+        String output = "    ____________________________________________________________\n"
+                + "     Hello! I'm Duke\n"
+                + "     What can I do for you?\n"
+                + "    ____________________________________________________________\n";
+        return output;
+    }
+
```
Comment (mukundrs) [inline src/main/java/duke/Ui.java]: Missing javadocs

**Review by mukundrs — COMMENTED** [review]
Sorry i messed up with the comments on my last review lololol. Here are my comments!

**Pull request closed without merging**